pkg/storageops/gce/gce.go

```diff
@@ -0,0 +1,540 @@
+type instance struct {
+	ID         string
+	Name       string
+	Hostname   string
+	Zone       string
+	Project    string
+	InternalIP string
+	ExternalIP string
+	LBRequest  string
+	ClientIP   string
+	Error      string
+}
+
+type assigner struct {
+	err error
+}
+
+func (a *assigner) assign(getVal func() (string, error)) string {
+	if a.err != nil {
+		return ""
+	}
+
+	s, err := getVal()
+	if err != nil {
+		a.err = err
+	}
+
+	return s
+}
+
+func getEnvValueStrict(key string) (val string, err error) {
+	if val = os.Getenv(key); len(val) != 0 {
+		return
+	}
+
+	err = fmt.Errorf("env variable %s is not set", key)
+	return
+}
+
+// gceInfo fetches the GCE instance metadata from the metadata server
+func gceInfo(inst *instance) {
+	a := &assigner{}
+	inst.ID = a.assign(metadata.InstanceID)
+	inst.Zone = a.assign(metadata.Zone)
+	inst.Name = a.assign(metadata.InstanceName)
+	inst.Hostname = a.assign(metadata.Hostname)
+	inst.Project = a.assign(metadata.ProjectID)
+	inst.InternalIP = a.assign(metadata.InternalIP)
+	inst.ExternalIP = a.assign(metadata.ExternalIP)
+
+	if a.err != nil {
+		inst.Error = a.err.Error()
+	}
+}
+
+func gceInfoFromEnv(inst *instance) {
+	a := &assigner{}
+	inst.Name = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_NAME") })
+	inst.Zone = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_ZONE") })
+	inst.Project = a.assign(func() (string, error) { return getEnvValueStrict("GCE_INSTANCE_PROJECT") })
+
+	if a.err != nil {
+		inst.Error = a.err.Error()
+	}
+}
+
+// IsDevMode checks if the pkg is invoked in developer mode where GCE credentials
+// are set as env variables
+func IsDevMode() bool {
+	var i = new(instance)
+	gceInfoFromEnv(i)
+	return len(i.Error) == 0
+}
+
+// NewClient creates a new GCE operations client
+func NewClient() (storageops.Ops, error) {
+	var i = new(instance)
+	if metadata.OnGCE() {
+		gceInfo(i)
+	} else if ok := IsDevMode(); ok {
+		gceInfoFromEnv(i)
+	} else {
+		return nil, fmt.Errorf("instance is not running on GCE")
+	}
+
+	if len(i.Error) != 0 {
+		return nil, fmt.Errorf("error while fetching instance information. Err: %v", i.Error)
+	}
+
+	c, err := google.DefaultClient(context.Background(), compute.ComputeScope)
+	if err != nil {
+		return nil, fmt.Errorf("failed to authenticate with google api. Err: %v", err)
+	}
+
+	service, err := compute.New(c)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create Compute service: %v", err)
+	}
+
+	return &gceOps{
+		inst:    i,
+		service: service,
+	}, nil
+}
+
+func (s *gceOps) waitStatus(id string, desired string) error {
+	actual := ""
+	for retries, maxRetries := 0, storageops.ProviderOpsMaxRetries; actual != desired && retries < maxRetries; retries++ {
+		d, err := s.service.Disks.Get(s.inst.Project, s.inst.Zone, id).Do()
+		if err != nil {
+			if ignore := notFoundRegex.MatchString(err.Error()); ignore {
+				time.Sleep(storageops.ProviderOpsRetryInterval)
```

Comment:
Since this is the same as 162, you can try to combine the logic. One way is to use `goto` , but I am sure there are others.